Dockerfile

```diff
@@ -1,109 +1,146 @@
         "https://packages.microsoft.com/keys/microsoft.asc" \
-        /usr/share/keyrings/microsoft.asc && \
-    echo "deb [signed-by=/usr/share/keyrings/deadsnakes.asc] https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu noble main" \
-      > /etc/apt/sources.list.d/deadsnakes-ppa.list && \
-    /usr/lib/apt/apt-helper download-file \
-        'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0xf23c5a6cf475977595c89f51ba6932366a755776' \
-        /usr/share/keyrings/deadsnakes.asc
-
+        /usr/share/keyrings/microsoft.asc
 
-# Install root dependencies, including Python
-COPY dependencies.txt /root/dependencies.txt
-# use space efficient utility from base image
+# install any additional system dependencies
+# NOTE: ensure .python-version is not excluded in .dockerignore
 RUN --mount=type=cache,target=/var/cache/apt \
-    /usr/bin/env ACCEPT_EULA=Y /root/docker-apt-install.sh /root/dependencies.txt
-
-#################################################
+    --mount=type=bind,source=.,target=/app \
+    /usr/bin/env ACCEPT_EULA=Y  /root/docker-apt-install.sh /app/dependencies.txt "python$(cat /app/.python-version)"
+
+# uv env var documentation: https://docs.astral.sh/uv/reference/environment/
+# copy files rather than symlink since the cache and the target are on different filesystems
+ENV UV_LINK_MODE=copy
+# compile at installation time, not import time
+ENV UV_COMPILE_BYTECODE=1
+# we are providing python via deadsnakes ppa
+# (as we require dynamic linking of openssl for security reasons)
+ENV UV_PYTHON_DOWNLOADS=never
+# set the directory for the venv
+ENV UV_PROJECT_ENVIRONMENT="/opt/venv"
+
+
+##################################################
+#
+# Build image
+#
+# Ok, now we have local base image with python and our system dependencies on.
+# We'll use this as the base for our builder image, where we'll build and
+# install any python packages needed.
 #
-# Next, use the dependencies image to create an image to build dependencies
-FROM ehrql-dependencies as ehrql-builder
+# We use a separate, disposable build image to avoid carrying the build
+# dependencies into the production image.
+FROM base-python as builder
 
-# install build time dependencies
-COPY build-dependencies.txt /root/build-dependencies.txt
-RUN /root/docker-apt-install.sh /root/build-dependencies.txt
+# Install any system build dependencies
+RUN --mount=type=cache,target=/var/cache/apt \
+    --mount=type=bind,source=dependencies-build.txt,target=/tmp/dependencies-build.txt \
+    /root/docker-apt-install.sh /tmp/dependencies-build.txt
+
+# Install everything in venv for isolation from system python libraries
+COPY --from=ghcr.io/astral-sh/uv:0.9 /uv /uvx /usr/local/bin/
 
+RUN uv venv
 
-# install everything in venv for isolation from system python libraries
-# hadolint ignore=DL3013,DL3042
+# The cache mount means a) /root/.cache is not in the image, and b) it's preserved
+# between docker builds locally, for faster dev rebuild.
```

Comment:
I think you're probably right that `uv` is so fast that, as long as we're preserving cached downloads between runs, it's fine to do away with the `pyproject.minimal.toml` workaround and just rebuild the environment every time anything changes.